File: backend/src/routes/recurring.ts
import express, { Response } from 'express';
import { docClient, TABLES, PutCommand, GetCommand, QueryCommand, UpdateCommand, DeleteCommand, generateId, getCurrentTimestamp } from '../db/dynamodb';
import { authenticateToken, AuthRequest } from '../middleware/auth';
import { RecurringCost } from '../types';

const router = express.Router();

// Get all recurring costs for user
router.get('/', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const result = await docClient.send(new QueryCommand({
      TableName: TABLES.RECURRING_COSTS,
      KeyConditionExpression: 'user_id = :userId',
      ExpressionAttributeValues: {
        ':userId': req.userId,
      },
    }));

    const recurringCosts = (result.Items || []) as RecurringCost[];
    // Sort by created_at descending
    recurringCosts.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());

    res.json(recurringCosts);
  } catch (error) {
    console.error('Get recurring costs error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Create recurring cost
router.post('/', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const { name, amount, frequency, category, start_date } = req.body;

    if (!name || !amount || !frequency || !category || !start_date) {
      return res.status(400).json({ error: 'All fields are required' });
    }

    if (frequency !== 'monthly' && frequency !== 'annual') {
      return res.status(400).json({ error: 'Frequency must be monthly or annual' });
    }

    const recurringId = generateId();
    const timestamp = getCurrentTimestamp();

    const recurringCost: RecurringCost = {
      id: recurringId,
      user_id: req.userId!,
      name,
      amount,
      frequency,
      category,
      start_date,
      created_at: timestamp,
    };

    await docClient.send(new PutCommand({
      TableName: TABLES.RECURRING_COSTS,
      Item: recurringCost,
    }));

    res.status(201).json(recurringCost);
  } catch (error) {
    console.error('Create recurring cost error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Update recurring cost
router.put('/:id', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;
    const { name, amount, frequency, category, start_date } = req.body;

    if (frequency && frequency !== 'monthly' && frequency !== 'annual') {
      return res.status(400).json({ error: 'Frequency must be monthly or annual' });
    }

    // Verify recurring cost belongs to user
    const result = await docClient.send(new GetCommand({
      TableName: TABLES.RECURRING_COSTS,
      Key: {
        user_id: req.userId,
        id: parseInt(id),
      },
    }));

    if (!result.Item) {
      return res.status(404).json({ error: 'Recurring cost not found' });
    }

    const updatedRecurringCost = await docClient.send(new UpdateCommand({
      TableName: TABLES.RECURRING_COSTS,
      Key: {
        user_id: req.userId,
        id: parseInt(id),
      },
      UpdateExpression: 'SET #name = :name, amount = :amount, frequency = :frequency, category = :category, start_date = :start_date',
      ExpressionAttributeNames: {
        '#name': 'name', // 'name' is a reserved keyword in DynamoDB
      },
      ExpressionAttributeValues: {
        ':name': name,
        ':amount': amount,
        ':frequency': frequency,
        ':category': category,
        ':start_date': start_date,
      },
      ReturnValues: 'ALL_NEW',
    }));

    res.json(updatedRecurringCost.Attributes);
  } catch (error) {
    console.error('Update recurring cost error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete recurring cost
router.delete('/:id', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;

    // Verify recurring cost belongs to user
    const result = await docClient.send(new GetCommand({
      TableName: TABLES.RECURRING_COSTS,
      Key: {
        user_id: req.userId,
        id: parseInt(id),
      },
    }));

    if (!result.Item) {
      return res.status(404).json({ error: 'Recurring cost not found' });
    }

    await docClient.send(new DeleteCommand({
      TableName: TABLES.RECURRING_COSTS,
      Key: {
        user_id: req.userId,
        id: parseInt(id),
      },
    }));

    res.json({ message: 'Recurring cost deleted successfully' });
  } catch (error) {
    console.error('Delete recurring cost error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/api/recurring.py
from fastapi import APIRouter, HTTPException, Depends
from typing import List
from .models import RecurringCostCreate, RecurringCostUpdate, RecurringCost
from .database import (
    get_recurring_costs_by_user,
    get_recurring_cost,
    create_recurring_cost,
    update_recurring_cost,
    delete_recurring_cost,
    generate_id,
    get_current_timestamp
)
from .middleware import get_current_user

router = APIRouter(prefix="/recurring", tags=["recurring"])


@router.get("/", response_model=List[RecurringCost])
async def get_all_recurring_costs(current_user: dict = Depends(get_current_user)):
    """Get all recurring costs for the authenticated user"""
    user_id = current_user['user_id']
    recurring_costs = get_recurring_costs_by_user(user_id)
    return recurring_costs


@router.get("/{recurring_id}", response_model=RecurringCost)
async def get_recurring_cost_by_id(recurring_id: int, current_user: dict = Depends(get_current_user)):
    """Get a specific recurring cost"""
    user_id = current_user['user_id']
    recurring_cost = get_recurring_cost(user_id, recurring_id)

    if not recurring_cost:
        raise HTTPException(status_code=404, detail="Recurring cost not found")

    return recurring_cost


@router.post("/", response_model=RecurringCost)
async def create_new_recurring_cost(
    recurring_data: RecurringCostCreate,
    current_user: dict = Depends(get_current_user)
):
    """Create a new recurring cost"""
    user_id = current_user['user_id']

    recurring = {
        'id': generate_id(),
        'user_id': user_id,
        'name': recurring_data.name,
        'amount': recurring_data.amount,
        'category': recurring_data.category,
        'frequency': recurring_data.frequency,
        'created_at': get_current_timestamp()
    }

    return create_recurring_cost(recurring)


@router.put("/{recurring_id}", response_model=RecurringCost)
async def update_recurring_cost_by_id(
    recurring_id: int,
    recurring_data: RecurringCostUpdate,
    current_user: dict = Depends(get_current_user)
):
    """Update a recurring cost"""
    user_id = current_user['user_id']

    # Check if recurring cost exists
    existing_recurring = get_recurring_cost(user_id, recurring_id)
    if not existing_recurring:
        raise HTTPException(status_code=404, detail="Recurring cost not found")

    # Build updates dict
    updates = {}
    if recurring_data.name is not None:
        updates['name'] = recurring_data.name
    if recurring_data.amount is not None:
        updates['amount'] = recurring_data.amount
    if recurring_data.category is not None:
        updates['category'] = recurring_data.category
    if recurring_data.frequency is not None:
        updates['frequency'] = recurring_data.frequency

    if not updates:
        raise HTTPException(status_code=400, detail="No updates provided")

    return update_recurring_cost(user_id, recurring_id, updates)


@router.delete("/{recurring_id}")
async def delete_recurring_cost_by_id(recurring_id: int, current_user: dict = Depends(get_current_user)):
    """Delete a recurring cost"""
    user_id = current_user['user_id']

    # Check if recurring cost exists
    existing_recurring = get_recurring_cost(user_id, recurring_id)
    if not existing_recurring:
        raise HTTPException(status_code=404, detail="Recurring cost not found")

    delete_recurring_cost(user_id, recurring_id)

    return {"message": "Recurring cost deleted successfully"}
